### ISSUE TITLE
How to tag information to point(s)

### ISSUE THREAD

**Issue by duongduong24** [potree/potree#1025]
Hi,

I am a civil engineer. Currently, I am researching point cloud data application for building facilities management. Thank you guys for the great works.

I am looking for a solution to tag information on a point or range of points. This means that I can click on point(s) and assign some information (such as name of object, id number, status,... etc) to that point(s).

Can I do this on Potree?

Thank you very much.

**Comment by HeadLessHUN** [comment]
hello do you mean annotation tags? you can do that and save it as a json file and load it back later if you want to save it. http://potree.org/potree/examples/annotations.html

**Comment by midnight-dev** [comment]
I was thinking about annotations, too, but maybe they mean something like custom attributes? Something akin to user-data? I don't have any experience with it, so I'm not sure if there is a practical way of merging extra data with a LAS before (or during) conversion.

Still, if you don't mind placing them manually (assuming it's a reasonable number), annotations seem like a good choice, especially for title:description text context.

**Comment by HeadLessHUN** [comment]
with pdal you can add any arbitrary attribute but to display that in potree it is complicated. 

if we are talking about point ranges as the  OP asked then i'm  thinking like a simple box volume as well for example with given names. maybe it could be a simple function to display the name in the box volume like the measurement. and it will highlight the clouds also. But this is just enduser solution if we are talking more like an automated import of objects from ifc or other files then the base threejs would be the superior solution but it would require an ifc loader which are not very common.

**Comment by smcavoy12** [comment]
Related: We're looking at a way to crowdsource noise removal on sonar data.

The system we're looking at will have potree version of the sonar in parralel with a repository of LAZ tiles. We'll tag points and volumes with the Potree web interface, then export the scene as a json using the "export potree" feature already built in. We'll pass those coordinates through PDAL to tag a new version of the LAZ files, which will then be reconverted into a potree format.

Markus had once mentioned that it'd be possible to just join a separate database up to the pointcloud, so you'd query a coordinate and return these values not baked directly into the potree pointcloud, but rather from an attached SQL database. I haven't seen this implemented yet.

It'd also be nice to have a sort of paintbrush tool to tag points, instead of doing by cubes, polygons, individually, that's farther down our list to investigate.

**Comment by midnight-dev** [comment]
You're right. I missed the part about ranges. It's not exactly ideal, but classification can help with this sort of thing, though it can also require a bunch of manual work depending on how you use it. ML is great for auto classifying ground, roads, buildings, etc. though manual classification of *which* building can be done, too. Other than that, the point selector in the measurements/tools menu would be able to display info about individual points like source ID and return number, and those values can be visualized with different colors/brightness in the Scene menu after selecting the point cloud. Have you tried this already, @duongduong24 to see if it fits your use case? It's not click & assign like annotations, but if you process the point cloud with something like CloudCompare or PDAL, you can modify clouds before being converted & viewed with potree.

**Comment by HeadLessHUN** [comment]
i think the Op wants to skip the big step of modeling objects and just tag points as a BIM solution for example add id,name, type to faucets, ducts, desks, etc... and later maybe get a report of the placed elements. But it is just my opinion.

**Comment by duongduong24** [comment]
> i think the Op wants to skip the big step of modeling objects and just tag points as a BIM solution for example add id,name, type to faucets, ducts, desks, etc... and later maybe get a report of the placed elements. But it is just my opinion.

Yes. Thats exactly what I'm thinking. Thanks

**Comment by duongduong24** [comment]
> You're right. I missed the part about ranges. It's not exactly ideal, but classification can help with this sort of thing, though it can also require a bunch of manual work depending on how you use it. ML is great for auto classifying ground, roads, buildings, etc. though manual classification of _which_ building can be done, too. Other than that, the point selector in the measurements/tools menu would be able to display info about individual points like source ID and return number, and those values can be visualized with different colors/brightness in the Scene menu after selecting the point cloud. Have you tried this already, @duongduong24 to see if it fits your use case? It's not click & assign like annotations, but if you process the point cloud with something like CloudCompare or PDAL, you can modify clouds before being converted & viewed with potree.

Thanks. Thats a grate idea. I will try it for my case. To be honest, I dont know much about programming so it will take me a while to learn how to do it.  
I want it to be a facility management system. That means I can use it to update, retrieve information (properties of assets, which are represented as point clouds) throughout the life of the building.
The annotation is great, but it's not flexible and convenience for the end-user as well as the huge amount of facility information.
Just like what @HeadLessHUN said. I want to skip the very big step of modeling objects (from point cloud data). I want to use Point Coud instead of BIM model.

**Comment by HeadLessHUN** [comment]
i think potree in current state cannot work fluently as a bim model.
As  i said the best method is the annotation tag for points and maybe my second thought which was the box volume with given name but you cannot assign more than 1 info on the boxes by default.

The idea is good but point clouds will affect performance and you cannot modify it. it will not replace a model especially in huge buildings.

the classification is an another story i don't think so you can use that because that way you will need to know the coordinates before converting to potree.

**Comment by duongduong24** [comment]
> i think potree in current state cannot work fluently as a bim model.
> As i said the best method is the annotation tag for points and maybe my second thought which was the box volume with given name but you cannot assign more than 1 info on the boxes by default.
> 
> The idea is good but point clouds will affect performance and you cannot modify it. it will not replace a model especially in huge buildings.
> 
> the classification is an another story i don't think so you can use that because that way you will need to know the coordinates before converting to potree.

yeah. So I think the Annotation is the best solution for my case. But do you know how to highlight a range of point cloud when you click on the annotation button. For example in the annotation example in the link, when you click on the annotation button Tree, the "tree" will be highlighted.

**Comment by HeadLessHUN** [comment]
it is not that easy. highlight without programming will require box volumes. (measure box). You can disable/enable those from the placed annotation. but it will be pain in the ass to place. unless you make a gui to place box and annotation and link them together.

with programming i would create a new box volume  class  (similar to the measurement boxes) and add a gui to place and fill it up with information.  this way you can place bounding box of the items and highlight them by glsl code. But this will require somebody who is not a beginner programmer. And it will be somewhat hard to place and to save it to file/project you need to modify that part as well to support the new type of object.
